Bluetooth: SMP_BR: Avoid to derive LTK from invalid BR LK

lylezhu2012 opened the pull request:
The LTK should not be generated from BR link key if there is a LTK exists and BR LK is weaker.

Do not drive LTK from BR LK in the case.

lylezhu2012 committed:
Bluetooth: SMP_BR: Avoid to derive LTK from invalid BR LK

The LTK should not be generated from BR link key if there is a LTK
exists and BR LK is weaker.

Do not drive LTK from BR LK in the case.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -1133,12 +1133,48 @@ static void smp_br_distribute_keys(struct bt_smp_br *smp)
+	addr.type = BT_ADDR_LE_PUBLIC;
+	bt_addr_copy(&addr.a, &conn->br.dst);
+
+	for (uint8_t id = 0; id < CONFIG_BT_ID_MAX; id++) {
```
Since BR/EDR doesn't support multiple identities, it's assumed that it always uses identity `0`. As such, this for-loop should be unnecessary.

lylezhu2012 replied:
The scenario is: 
1. The LK of BR is derived from BLE LTK. 
2. Then disconnect from LE ACL connection. 
3. And build the ACL connection on BR transport. 
4. After the connection is established, the peer force start the pairing procedure. 
5. After the pairing procedure of BR is done, the peer tries to derive LTK from LK.
According to sepc, `If an LE LTK already exists and the BR/EDR link key is weaker in either strength or MITM protection, then neither device shall generate an LE LTK using cross-transport key derivation from a BR/EDR link key.`, we need to find whether there is a LTK has been boned.
But in the step one, the id of the LE key is unknown (the connection has been broken in step 5). So in order to avoid LTK be overwrote, I use for-loop to search LE keys in all ids.

jhedberg replied:
The way the multi-identity feature is designed (or at least intended), is that id 0 is what pairs up with the BR/EDR identity, in case of a dual-mode device. The other identities should be treated as single-mode (LE-only) and no cross-transport key derivation should be performed for them.

lylezhu2012 replied:
Got it. Let me update the PR.

lylezhu2012 replied:
Sorry, I have checked that there is not any limitation that only `BT_ID_DEFAULT` can be used to derive LK. I think we need to avoid LK be derived if the id is not BT_ID_DEFAULT.

lylezhu2012 replied:
> Got it. Let me update the PR.

Updated.

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -1133,12 +1133,48 @@ static void smp_br_distribute_keys(struct bt_smp_br *smp)
+		return false;
+	}
+
+	if (conn->encrypt == 0x02) {
```
What's the `0x02` magic number? Please add a define for it (perhaps to `hci_types.h`?). Probably best to add defines for `0x00` and `0x01` as well in the same go.

jhedberg replied:
Note: I can see that the old code doesn't have it either, but it would be nice to improve the readability in the same go when you're modifying the code.

lylezhu2012 replied:
yes. I will add a macro for it.

lylezhu2012 replied:
Updated.

lylezhu2012 committed:
Bluetooth: SMP_BR: Use macros instead of hard code

Define macros for encryption value,
#define BT_HCI_ENCRYPTION_OFF           0x00
#define BT_HCI_ENCRYPTION_ON_LE_AES_CCM 0x01
#define BT_HCI_ENCRYPTION_ON_BR_E0      0x01
#define BT_HCI_ENCRYPTION_ON_BR_AES_CCM 0x02

Use the macros to replace the hard code.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on subsys/bluetooth/host/smp.c:
```diff
@@ -1133,12 +1133,48 @@ static void smp_br_distribute_keys(struct bt_smp_br *smp)
+	/**
+	 * Core v6.0, Vol 3, Part C, 14.1 Cross-transport key derivation
+	 *
+	 * If an LE LTK already exists and the BR/EDR link key is weaker in either strength
```
It has been shown in the code, `key->flags & BT_LINK_KEY_AUTHENTICATED` and `conn->encrypt == 0x02`.

MarkWangChinese replied:
Do you mean that: `!(key->flags & BT_LINK_KEY_AUTHENTICATED)` indicates `BR/EDR link key has no MITM protection` and `conn->encrypt != 0x02` indicates `BR/EDR link key strength is weaker than current LTK`? Do we need any comparison to check the `strength`.?

lylezhu2012 replied:
In current implementation, I only found no authenticated issue. I don't think there are other issues here. For strength key check, I think it is also has been checked.
